### DISCUSSION TITLE
GTA 5 Broken Modkit Bones

### DISCUSSION THREAD

**Discussion by xesdoog** [YimMenu/YimMenu#3269] [Research]
Hello yimmies,

A while ago I borrowed a small script from [Harmless](https://github.com/harmless05) and started adding more features to it. It now has NOS and some other car-related stuff. The Speed Boost (*or NOS*) option tries to simulate nitrous by giving you more torque and increasing your top speed (it's not a horn boost) and as a nice addition, it spawns a flame on your exhaust:

![civic](https://github.com/YimMenu/YimMenu/assets/66764345/4d00e0ee-29e2-44d3-bf57-7091c88d91f9)

This is where things start to go south. The flames only show on stock exhausts and maybe 3 or 4 modkit exhausts. After doing some research, I found an [old FiveM issue](https://github.com/citizenfx/fivem/issues/2003) about this exact thing and it turns out R* broke some modkits in b3095.

There is also a fix for it which could be implemented in YimMenu but C++ is out of my league I wouldn't even know where to start so I'm opening this discussion hoping that a kind soul could turn [this fix](https://github.com/citizenfx/fivem/commit/730bb5abb84effab241604e2889b60a556fec728) into a YimMenu PR.

Thanks ♥

**Comment by xiaoxiao921** [comment]
The fiveM issue doesn't say this at all?
It says that b3095 introduced new natives that fixed the mentioned issue, and that the fix you are talking about is backporting it to older versions of the game

**Reply by xesdoog** [reply]
> The fiveM issue doesn't say this at all?
> It says that b3095 introduced new natives that fixed the mentioned issue, and that the fix you are talking about is backporting it to older versions of the game

Hello xiaoxiao,

I'm on mobile now but I'll try my best to make things clear

3095 didn't fix anything, in fact it introduced this issue and to double check that it's still present, [here's the script](https://github.com/xesdoog/TokyoDrift) give it a go, activate Speed Boost and PTFX then accelerate and press left shift while having a custom exhaust then go to LSC or use YimMenu to change the exhaust and you'll see that the particle effect will only spawn on stock exhausts. 

The FiveM fix is doing this:

![Screenshot_20240620-182518_GitHub.jpg](https://github.com/YimMenu/YimMenu/assets/66764345/57922580-04fb-4cde-9c62-323aa44a0df2)

Although I don't know anything about hooking and stuff but I'm sure you understand what this comment is talking about

**Comment by xiaoxiao921** [comment]
Please reread the fivem issue, all the commit you linked does is doing extra code if the version is INFERIOR to 3095!
all the guy says is that 3095 added extra vehicle natives for this particular issue, maybe you just need to use those instead?
![image](https://github.com/YimMenu/YimMenu/assets/837334/72291320-3d3b-47f6-be12-10a9bb2209f8)

**Reply by Deadlineem** [reply]
> Please reread the fivem issue, all the commit you linked does is doing extra code if the version is INFERIOR to 3095! all the guy says is that 3095 added extra vehicle natives for this particular issue, maybe you just need to use those instead? ![image](https://private-user-images.githubusercontent.com/837334/341521086-72291320-3d3b-47f6-be12-10a9bb2209f8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg5MTA4NTcsIm5iZiI6MTcxODkxMDU1NywicGF0aCI6Ii84MzczMzQvMzQxNTIxMDg2LTcyMjkxMzIwLTNkM2ItNDdmNi1iZTEyLTEwYTliYjIyMDlmOC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjIwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYyMFQxOTA5MTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kMjVkOGY3M2YzYWNjZTllOGI5MjdiZmU4MzA1Mzg4MDE4MWZjMDBiMGE3MmVkZjg1ZTQ3YzczOWM0OTdmZDEwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.t_qRv5Nf2hUFfAoWOx2w4xFcNoxrAO32iYXUEzYCKzY)

We were not aware they added new natives for this such thing, but this is why we are here @xiaoxiao921 and we appreciate your input on the subject, you've helped us figure this out, we have been stuck on this for a few days trying to figure out why tf the modkit bones wouldnt use the ptfx but the stock exhaust would.

**Reply by Deadlineem** [reply]
attempt to call a nil value (field '_GET_VEHICLE_EXHAUST_BONE')
attempt to call a nil value (field '_GET_VEHICLE_MAX_EXHAUST_BONE_COUNT')

Even though these native bindings are defined, it returns this error when using through lua @xiaoxiao921

**Comment by xiaoxiao921** [comment]
I just checked your code and you are indeed not using them

**Reply by xesdoog** [reply]
Sorry about the confusion you're right about the PR but yes I'm not using those natives because they're not included in the current Lua binding

**Reply by Deadlineem** [reply]
> Sorry about the confusion you're right about the PR but yes I'm not using those natives because they're not included in the current Lua binding

https://github.com/YimMenu/YimMenu/blob/2f79436d8efb7f673b46b39c4e667b5a9f4fe3f3/src/lua/natives/lua_native_binding_VEHICLE.cpp#L4183C2-L4198C1

it says they are in there but no idea if they work or not, haven't tested.

**Reply by xiaoxiao921** [reply]
in lua
VEHICLE.GET_VEHICLE_EXHAUST_BONE_
should work

**Reply by Deadlineem** [reply]
> in lua VEHICLE.GET_VEHICLE_EXHAUST_BONE_ should work

AAAHH i see VEHICLE.GET_VEHICLE_MAX_EXHAUST_BONE_COUNT_ with the trailing underscore instead of with the prefix underscore.

**Reply by xesdoog** [reply]
> > Sorry about the confusion you're right about the PR but yes I'm not using those natives because they're not included in the current Lua binding
> 
> https://github.com/YimMenu/YimMenu/blob/2f79436d8efb7f673b46b39c4e667b5a9f4fe3f3/src/lua/natives/lua_native_binding_VEHICLE.cpp#L4183C2-L4198C1
> 
> it says they are in there but no idea if they work or not, haven't tested.

@Deadlineem Can I go back in time and un-embarrass myself? 😭
@xiaoxiao921 I'm so sorry dude 💀 both natives work I'm just an idiot

**Comment by Deadlineem** [comment]
@xiaoxiao921 I tried iterating over the bone count and applying it to the exhausts to no avail.  

VEHICLE.GET_VEHICLE_EXHAUST_BONE_ only returns a boolean value (True/False) and nothing useful to apply ptfx to modified exhaust tips.

When i use the bone count, it returns "32" in a gui.show_message() which i thought would be the bone ID (not index, right?) and tried to apply the ptfx on it that way but nothing.

the desired outcome is to apply the veh_nitrous particle effect to the exhaust bones on each vehicle, which currently only applies to stock exhaust (exhaust through exhaust_16), but does not apply to most modified exhausts, like the titanium tip for the infernus classic for example.

Is there something I am doing wrong?

**Reply by xiaoxiao921** [reply]
local someBool, boneIndex, axisX = VEHICLE.GET_VEHICLE_EXHAUST_BONE_(veh, index, 0, 0)
should work I think

**Reply by xesdoog** [reply]
Used this and it works
```lua
local exhaustCount = VEHICLE.GET_VEHICLE_MAX_EXHAUST_BONE_COUNT_()
for i = 0, exhaustCount do
      local retBool, boneIndex = VEHICLE.GET_VEHICLE_EXHAUST_BONE_(current_vehicle, i, retBool, boneIndex)
      if retBool then
          GRAPHICS.USE_PARTICLE_FX_ASSET(effect)
          nosPtfx = GRAPHICS.START_NETWORKED_PARTICLE_FX_LOOPED_ON_ENTITY_BONE("veh_nitrous", current_vehicle, ptfx_X_axis, ptfx_Y_axis, 0.0, 0.0, 0.0, 0.0, boneIndex, 1.0, false, false, false, 0, 0, 0)
           table.insert(nosptfx_t, nosPtfx)
           nos_started = true
        end
  end
  ```
  Thank you for the help xiaoxiao ♥
  You too @Deadlineem WE DID IT! 😁
